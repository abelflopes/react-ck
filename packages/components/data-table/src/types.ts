export const sortModes = ["asc", "desc", "none"] as const;

export type SortMode = (typeof sortModes)[number];

/** Type representing the data structure for the DataTable component  */
export type TableData = Array<Record<string, React.ReactNode>>;

export type SortCallback<T extends TableData> = (key: keyof T[number]) => void;
