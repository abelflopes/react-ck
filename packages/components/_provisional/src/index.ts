export * from "./image";

export * from "./avatar";

export * from "./dropdown";

export * from "./dock";

export * from "./breadcrumbs";

export * from "./scrollable-container";

export * from "./position-engine";

export * from "./top-bar";

export * from "./snackbar";

export * from "./overlay-spinner";

export * from "./menu";

export * from "./tooltip";

export * from "./popover";

export * from "./chat";

export * from "./file-uploader";

export * from "./attachment";

export * from "./snippet";
