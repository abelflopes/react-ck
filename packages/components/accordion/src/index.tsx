import React from "react";
import { List, type ListProps } from "@react-ck/list";
import { Collapse } from "@react-ck/collapse";
import styles from "./styles/index.module.scss";

export interface AccordionItem {
  header: NonNullable<React.ReactNode>;
  children: NonNullable<React.ReactNode>;
}

export interface AccordionProps extends Omit<ListProps, "items"> {
  items: AccordionItem[];
}

/**
 * An accordion is a vertically stacked list of interactive items. Each item can be "collapsed" with only a summary visible or “expanded” to show the full content for that item.
 * @param props - {@link AccordionProps}
 * @returns a React element
 */

export const Accordion = ({
  items,
  ...otherProps
}: Readonly<AccordionProps>): React.ReactElement => (
  <List
    {...otherProps}
    items={items.map(({ header, children }) => (
      <Collapse key={String(JSON.stringify(header))} header={header}>
        <div className={styles.content}>{children}</div>
      </Collapse>
    ))}
  />
);
