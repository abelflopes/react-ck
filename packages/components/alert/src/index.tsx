import React from "react";
import classNames from "classnames";
import styles from "./styles/index.module.scss";
import { Text } from "@react-ck/text";

export interface AlertProps extends React.HTMLAttributes<HTMLDivElement> {
  heading?: string;
  /** Specifies the visual style of the alert  */
  skin?: "neutral" | "primary" | "negative" | "average" | "positive";
}

// TODO: add dismiss

/**
 * Alert is a short message that provides contextual feedback in a prominent way.
 * @param props - {@link AlertProps}
 * @returns a React element
 */

export const Alert = ({
  heading,
  skin = "neutral",
  children,
  className,
  ...otherProps
}: Readonly<AlertProps>): React.ReactElement => (
  <div {...otherProps} className={classNames(styles.root, styles[skin], className)}>
    {heading && (
      <Text type="h4" as="p" variation="bold" margin={false} className={styles.heading}>
        {heading}
      </Text>
    )}
    {children}
  </div>
);
