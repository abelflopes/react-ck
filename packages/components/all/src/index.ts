export * from "./text";

export * from "./divider";

export * from "./banner";

export * from "./card";

export * from "./container";

export * from "./chip";

export * from "./collapse";

export * from "./list";

export * from "./overlay";
