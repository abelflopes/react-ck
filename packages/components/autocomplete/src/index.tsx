import React, { useEffect, useId, useMemo } from "react";
import styles from "./styles/index.module.scss";
import classNames from "classnames";
import { useFormFieldContext, type FormFieldProps } from "@react-ck/form-field";
import { AutocompleteOption } from "./AutocompleteOption";

interface AutocompleteProps extends React.InputHTMLAttributes<HTMLInputElement> {
  skin?: FormFieldProps["skin"];
}

/**
 * Autocomplete is a text input that proactively suggests options based on the userâ€™s initial input. The suggested options appear alongside the input in a panel.
 * @param props - {@link AutocompleteProps}
 * @returns a React element
 */

const Autocomplete = ({
  skin,
  id,
  className,
  children,
  ...props
}: Readonly<AutocompleteProps>): React.ReactElement => {
  const listId = useId();
  const formFieldContext = useFormFieldContext();

  const computedSkin = useMemo(
    () => formFieldContext?.skin ?? skin ?? "default",
    [formFieldContext?.skin, skin],
  );

  const computedId = useMemo(() => formFieldContext?.id ?? id, [formFieldContext?.id, id]);

  // Validate usage inside form field
  useEffect(() => {
    // Is not inside form field, skip
    if (formFieldContext === undefined) return;

    // Is inside form field
    if (skin)
      throw new Error("When using autocomplete inside form field, define skin on the form field");
    else if (id)
      throw new Error("When using autocomplete inside form field, define id on the form field");
  }, [formFieldContext, id, skin]);

  return (
    <>
      <input
        {...props}
        id={computedId}
        list={listId}
        className={classNames(
          styles.root,
          formFieldContext === undefined && styles.standalone,
          className,
          styles[`skin_${computedSkin}`],
        )}
      />
      <datalist id={listId}>{children}</datalist>
    </>
  );
};

Autocomplete.Option = AutocompleteOption;

export { Autocomplete, type AutocompleteProps };
