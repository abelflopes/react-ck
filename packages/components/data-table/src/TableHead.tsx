import React from "react";
import classNames from "classnames";
import { type SortMode } from "./types";
import { Icon } from "@react-ck/icon";
import { IconChevronUp } from "@react-ck/icon/icons/IconChevronUp";
import { IconChevronDown } from "@react-ck/icon/icons/IconChevronDown";
import * as styles from "./styles/index.module.scss";

export interface TableHeadProps extends React.TdHTMLAttributes<HTMLTableCellElement> {
  sorting: SortMode | null;
}

export const TableHead = ({
  sorting,
  className,
  children,
  ...otherProps
}: Readonly<TableHeadProps>): React.ReactElement => (
  <th
    {...otherProps}
    className={classNames({
      className,
      [`${styles.sortable_header}`]: sorting,
    })}>
    {children}

    {sorting === "asc" && (
      <Icon className={styles.sortable_header_icon}>
        <IconChevronUp />
      </Icon>
    )}

    {sorting === "desc" && (
      <Icon className={styles.sortable_header_icon}>
        <IconChevronDown />
      </Icon>
    )}
  </th>
);
