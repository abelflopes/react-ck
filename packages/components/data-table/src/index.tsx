import React, { useCallback, useMemo, useState } from "react";
import { Table, type TableProps } from "@react-ck/table";
import * as CC from "change-case";
import { stringFromObject } from "./utils/string-from-object";
import { Icon } from "@react-ck/icon";
import styles from "./styles/index.module.scss";
import { componentToText } from "./utils/component-to-text";

// TODO: add pagination
// TODO: add section table

const sortModes = ["asc", "desc", "none"] as const;

/** Type representing the data structure for the DataTable component  */
type TableData = Array<Record<string, React.ReactNode>>;

export type SortCallback<T extends TableData> = (key: keyof T[number]) => void;

/**
 * DataTableProps interface represents the props for the DataTable component.
 * @typeParam T - The type of data provided for the DataTable.
 */
export interface DataTableProps<T extends TableData> extends Omit<TableProps, "children"> {
  /** Headers mapping to define column names and their corresponding header content */
  headers?: Partial<Record<keyof T[number], React.ReactNode>>;
  /** Data to be displayed in the table. Should be an array of objects with keys matching the headers' keys (non-mandatory) */
  data: T;
  /** Automatically create table headers based on the row keys */
  autoHeaders?: boolean;
  /** Allow user to sort tale content by clicking on table headers */
  sortable?: boolean | Array<keyof T[number]>;
  /** Sort callback to allow handling sorting externally instead of automatic string / number sorting */
  onSort?: SortCallback<T>;
}

/**
 * Data table is a component that transforms a JSON structure into a table
 * @param props - {@link DataTableProps}
 * @returns a React element
 */

export const DataTable = <T extends TableData>({
  headers,
  data,
  autoHeaders,
  sortable,
  onSort,
  ...otherProps
}: Readonly<DataTableProps<T>>): React.ReactElement => {
  const [sortMode, setSortMode] = useState<(typeof sortModes)[number]>("none");
  const [sortKey, setSortKey] = useState<keyof T[number] | undefined>(undefined);

  const keys = useMemo(
    () =>
      [...data.flatMap((index) => Object.keys(index)), ...(headers ? Object.keys(headers) : [])]
        // Remove repeated
        .reduce<string[]>(
          (previous, current) => [...previous, ...(previous.includes(current) ? [] : [current])],
          [],
        ),
    [data, headers],
  );

  // Generate headers object merged into autoHeaders, if applicable
  const computedHeaders = useMemo(
    () => ({
      ...(autoHeaders ? Object.fromEntries(keys.map((k) => [k, CC.capitalCase(k)])) : undefined),
      ...headers,
    }),
    [autoHeaders, keys, headers],
  );

  // Handle sorting
  const defaultOnSort = useCallback<SortCallback<T>>(
    (key) => {
      if (!sortable || (Array.isArray(sortable) && !sortable.includes(key))) return;

      if (key === sortKey) {
        const nextSortMode = sortModes[(sortModes.indexOf(sortMode) + 1) % sortModes.length];
        if (!nextSortMode) throw new Error("Invalid sort mode");
        setSortMode(nextSortMode);
      } else {
        setSortMode("asc");
      }

      setSortKey(key);
    },
    [sortKey, sortMode, sortable],
  );

  // Sort data if applicable
  const sortedData = useMemo(() => {
    if (onSort || sortMode === "none" || !sortKey) return data;

    const k = String(sortKey);

    return [...data].sort((a, b) => {
      /* eslint-disable @typescript-eslint/no-non-null-assertion -- files always exist */
      let valueA = a[k]!;
      let valueB = b[k]!;
      /* eslint-enable */

      if ([valueA, valueB].every((i) => !Number.isNaN(Number(i)))) {
        // if both are numbers
        valueA = Number(valueA);
        valueB = Number(valueB);
      } else if ([valueA, valueB].every(React.isValidElement)) {
        // if both are react elements
        valueA = (componentToText(valueA) ?? "").trim().toLowerCase();
        valueB = (componentToText(valueB) ?? "").trim().toLowerCase();
      } else {
        // treat as regular strings
        valueA = String(valueA).trim().toLowerCase();
        valueB = String(valueB).trim().toLowerCase();
      }

      if (valueA > valueB) return sortMode === "desc" ? -1 : 1;
      else if (valueA < valueB) return sortMode === "desc" ? 1 : -1;
      return 0;
    });
  }, [data, onSort, sortKey, sortMode]);

  return (
    <Table {...otherProps}>
      {Object.keys(computedHeaders).length > 0 && (
        <thead>
          <tr>
            {keys.map((key) => (
              <th
                key={key}
                className={
                  sortable === true || (Array.isArray(sortable) && sortable.includes(key))
                    ? styles.sortable_header
                    : undefined
                }
                onClick={() => {
                  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- FIXME: remove type assertion
                  defaultOnSort(key as keyof T[number]);
                }}>
                {computedHeaders[key]}

                {sortable === true ||
                  (Array.isArray(sortable) && sortable.includes(key) && (
                    <>
                      {sortKey === key && sortMode === "asc" && (
                        <Icon name="chevron-up" className={styles.sortable_header_icon} />
                      )}

                      {sortKey === key && sortMode === "desc" && (
                        <Icon name="chevron-down" className={styles.sortable_header_icon} />
                      )}
                    </>
                  ))}
              </th>
            ))}
          </tr>
        </thead>
      )}

      <tbody>
        {sortedData.map((row) => (
          <tr key={stringFromObject(row)}>
            {keys.map((key) => (
              <td key={key}>{row[key]}</td>
            ))}
          </tr>
        ))}
      </tbody>
    </Table>
  );
};
