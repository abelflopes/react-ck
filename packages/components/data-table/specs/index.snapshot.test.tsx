import React from "react";
import { DataTable } from "../src/index";
import renderer from "react-test-renderer";

describe("Snapshot DataTable", () => {
  test("renders correctly", async () => {
    const data = [{ a: 1, b: 2, c: 3 }];
    const tree = renderer.create(<DataTable data={data} data-testid="table" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
