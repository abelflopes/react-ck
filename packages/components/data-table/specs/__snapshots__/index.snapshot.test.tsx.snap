// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot DataTable renders correctly 1`] = `
<div
  className="root"
  data-testid="table"
>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th />
        <th />
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          2
        </td>
        <td>
          3
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
